### PR TITLE
(stm32) drivers: firewall: Add firewall illegal access collectors drivers (IAC, SERC)

### PR DESCRIPTION
**Pull request by GseoC**
Add IAC and SERC drivers that allows OP-TEE to handle illegal accesses captured by the IAC and SERC peripheral. They ensure the robustness of the platform and ease comprehension of firewall exceptions

### CODE REVIEW

**Comment by GseoC** [comment]
@etienne-lms , can you review this P-R please?

**Review by etienne-lms — COMMENTED** [review]
Sorry I forgot to post them. Here they are, mainly minor issues.
I'll have another review round...

File: core/drivers/firewall/stm32_serc.c [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+	const void *fdt = serc_dev.fdt;
+	int node = serc_dev.node;
+
+	assert(serc_dev.fdt);
+
+	fdt_fill_device_info(fdt, &dt_info, node);
+	if (dt_info.status == DT_STATUS_DISABLED ||
+	    dt_info.reg == DT_INFO_INVALID_REG ||
+	    dt_info.clock == DT_INFO_INVALID_CLOCK ||
+	    dt_info.interrupt == DT_INFO_INVALID_INTERRUPT)
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/firewall/stm32_serc.c:125]: Applying same modification for the IAC driver

**Comment by GseoC** [comment]
Comments addressed

**Comment by GseoC** [comment]
@etienne-lms, seems good to you?

File: core/drivers/firewall/stm32_iac.c [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+		uint32_t offset = sizeof(uint32_t) * i;
+		uint8_t tries = 0;
+
+		isr = io_read32(base + _IAC_ISR0 + offset);
+		isr &= io_read32(base + _IAC_IER0 + offset);
+
+		if (!isr)
+			continue;
+
+		if (isr) {
```
Comment (etienne-lms) [inline core/drivers/firewall/stm32_iac.c:159]: could remove this test

File: core/drivers/firewall/stm32_iac.c [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+	struct iac_driver_data *ddata = iac_dev.ddata;
+	vaddr_t base = iac_dev.pdata.base;
+	unsigned int nreg = DIV_ROUND_UP(ddata->num_ilac, _PERIPH_IDS_PER_REG);
+	bool do_panic = false;
+	unsigned int i = 0;
+	uint32_t isr = 0;
+
+	for (i = 0; i < nreg; i++) {
+		uint32_t offset = sizeof(uint32_t) * i;
+		uint8_t tries = 0;
```
Comment (etienne-lms) [inline core/drivers/firewall/stm32_iac.c:151]: by the way, label `tries` is confusing. This is a simple bit counter, prefer name it `n` or `j` or like.
I think the loop at line 165 would be simpler to read as:
```c
for (n = 0; n < _PERIPH_IDS_PER_REG; n++) {
   ...
   if (!isr)
       break;
}

File: core/drivers/firewall/stm32_iac.c [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	    dt_info.reg == DT_INFO_INVALID_REG)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	pdata->irq_chip = interrupt_get_main_chip();
+	res = interrupt_dt_get(fdt, node, &pdata->irq_chip, &pdata->irq);
+	if (res)
+		return res;
+
+	base.pa = dt_info.reg;
+	pdata->base = io_pa_or_va_secure(&base, 1);
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/firewall/stm32_iac.c:135]: Applying first suggestion for consistency with other drivers

File: core/drivers/firewall/stm32_serc.c [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+static TEE_Result probe_serc_device(const void *fdt, int node)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = stm32_serc_parse_fdt(fdt, node);
+	if (res)
+		return res;
+
+	/* Asymmetric clock enable to have the SERC running */
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/firewall/stm32_serc.c:215]: Changing to "unbalanced" as I prefer to justify it is done on purpose

**Comment by GseoC** [comment]
Comments addressed, thanks

**Comment by GseoC** [comment]
Are you satisfied with the changes @etienne-lms ?

**Comment by etienne-lms** [comment]
Looks good to me; Could you squash the fixup commits, i'll give a last review round afterwards.

**Comment by GseoC** [comment]
Patches squashed. I took the liberty to remove the usage of an old local variable in the IAC driver to be aligned with the SERC driver

**Review by etienne-lms — COMMENTED** [review]
Few remaining minor comments for commits
"drivers: firewall: add stm32 IAC support" and
"drivers: firewall: add stm32 SERC support".

`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"dts: stm32: add IAC node in stm32mp251 SoC device tree file",
"dts: stm32: add SERC node in stm32mp251 SoC device tree file" and
"drivers: stm32_rif: add stm32_rif_access_violation_action()".

**Comment by GseoC** [comment]
Given the comments, I've not bothered making specific review commits when applying them.
Tags applied

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commmits
"drivers: firewall: add stm32 IAC support" (remaining comment addressed or not),
"plat-stm32mp2: default enable IAC on stm32mp2x platforms",
"drivers: firewall: add stm32 SERC support" and
"plat-stm32mp2: default enable SERC peripheral".

**Comment by GseoC** [comment]
Last comment addressed and tags applied, thanks

**Pull request merged**